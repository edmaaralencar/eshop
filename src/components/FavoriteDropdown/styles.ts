import styled from 'styled-components'

import * as HoverCard from '@radix-ui/react-hover-card'

export const Wrapper = styled(HoverCard.Portal)``

export const Content = styled(HoverCard.Content)`
  border-radius: 6px;
  width: 30rem;
  background-color: ${({ theme }) => theme.colors.white};
  box-shadow: hsl(206 22% 7% / 35%) 0px 10px 38px -10px,
    hsl(206 22% 7% / 20%) 0px 10px 20px -15px;

  animation-duration: 0.6s;
  animation-timing-function: cubic-bezier(0.16, 1, 0.3, 1);

  &[data-side='top'] {
    animation-name: slideUp;
  }
  &[data-side='bottom'] {
    animation-name: slideDown;
  }

  @keyframes slideUp {
    from {
      opacity: 0;
      transform: translateY(10px);
    }
    to {
      opacity: 1;
      transform: translateY(0);
    }
  }

  @keyframes slideDown {
    from {
      opacity: 0;
      transform: translateY(-10px);
    }
    to {
      opacity: 1;
      transform: translateY(0);
    }
  }
`

export const Arrow = styled(HoverCard.Arrow)`
  fill: white;
`

export const EmptyList = styled.div`
  display: grid;
  place-items: center;

  width: 100%;
  height: 8rem;

  p {
    font-size: 1.4rem;
  }
`

export const FavoritesList = styled.div`
  display: flex;
  flex-direction: column;

  max-height: 40rem;
  overflow: auto;
`

export const Favorite = styled.div`
  position: relative;

  display: flex;
  flex-direction: column;
  gap: 1.6rem;

  padding: 1.6rem;
`

export const FavoriteContainer = styled.div`
  display: flex;
  align-items: center;
  gap: 1.6rem;

  > div {
    display: flex;
    flex-direction: column;
    gap: 0.4rem;

    span {
      font-size: 1.4rem;
      color: ${({ theme }) => theme.colors.title};
    }

    strong {
      font-size: 1.8rem;
      color: ${({ theme }) => theme.colors.text.purple};
    }
  }
`

export const ButtonContainer = styled.div`
  position: absolute;
  top: 1rem;
  right: 4rem;

  button {
    position: absolute;

    width: 3.2rem;
    height: 3.2rem;
  }
`
