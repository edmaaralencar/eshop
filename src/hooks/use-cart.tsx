import { useQuery } from '@tanstack/react-query'
import {
  createContext,
  ReactNode,
  useContext,
  useEffect,
  useState
} from 'react'
import { IProduct } from '../@types/product'
import { api } from '../lib/axios'
import {
  getStorageItem,
  removeStorageItem,
  setStorageItem
} from 'utils/storage'

type CartItem = {
  id: string
  quantity: number
}

type CartProviderProps = {
  children: ReactNode
}

type CartContextData = {
  cartItems: Array<{
    product: IProduct
    quantity: number
  }>
  quantity: number
  total: number
  isLoading: boolean
  clearCart: () => void
  addProductToCart: (product: CartItem) => void
  handleIncrementProductInCart: (product: CartItem) => void
  handleDecrementProductInCart: (product: CartItem) => void
}

const CartContext = createContext({} as CartContextData)

export function CartProvider({ children }: CartProviderProps) {
  const [cartItems, setCartItems] = useState<CartItem[]>([])

  useEffect(() => {
    const cartItemsInStorage = getStorageItem('cart')

    if (cartItemsInStorage) {
      setCartItems(cartItemsInStorage)
    }
  }, [])

  const { data, isLoading } = useQuery(
    ['cart/items', cartItems],
    async () => {
      let url = `/products/items?`

      for (const item of cartItems) {
        url += `id=${item.id}&`
      }

      const { data } = await api.get(url)

      return data
    },
    {
      enabled: cartItems.length > 0
    }
  )

  const products = data?.map((product: IProduct) => ({
    product,
    quantity: cartItems.find((item) => item.id === product.id)?.quantity
  }))

  const total = products?.reduce(
    (acc: number, item: { product: IProduct; quantity: number }) =>
      (acc += item.product.price * item.quantity),
    0
  )

  function addProductToCart(product: CartItem) {
    setCartItems((prevState) => {
      const productIsInCart = prevState.find((item) => item.id === product.id)

      if (productIsInCart) {
        return prevState.map((item) =>
          item.id === product.id
            ? { ...product, quantity: product.quantity }
            : item
        )
      } else {
        return [...prevState, product]
      }
    })

    const cartItemsInStorage: CartItem[] = getStorageItem('cart')

    if (cartItemsInStorage) {
      const productIsInCart = cartItemsInStorage?.find(
        (item) => item.id === product.id
      )

      if (productIsInCart) {
        const productWithCorrectQuantity = cartItemsInStorage.map((item) =>
          item.id === product.id
            ? { ...product, quantity: product.quantity }
            : item
        )
        setStorageItem('cart', productWithCorrectQuantity)
      } else {
        setStorageItem('cart', [...cartItemsInStorage, product])
      }
    } else {
      setStorageItem('cart', [product])
    }
  }

  function handleIncrementProductInCart(product: CartItem) {
    setCartItems((prevState) => {
      return prevState.map((item) =>
        item.id === product.id
          ? { ...product, quantity: (product.quantity += 1) }
          : item
      )
    })

    const cartItemsInStorage: CartItem[] = getStorageItem('cart')

    const productIsInCart = cartItemsInStorage?.find(
      (item) => item.id === product.id
    )

    if (productIsInCart) {
      const productWithCorrectQuantity = cartItemsInStorage.map((item) =>
        item.id === product.id
          ? { ...productIsInCart, quantity: (productIsInCart.quantity += 1) }
          : item
      )
      setStorageItem('cart', productWithCorrectQuantity)
    } else {
      setStorageItem('cart', [...cartItemsInStorage, product])
    }
  }

  function handleDecrementProductInCart(product: CartItem) {
    setCartItems((prevState) => {
      if (product.quantity === 1) {
        return prevState.filter((item) => item.id !== product.id)
      }

      return prevState.map((item) =>
        item.id === product.id
          ? { ...product, quantity: (product.quantity -= 1) }
          : item
      )
    })

    const cartItemsInStorage: CartItem[] = getStorageItem('cart')

    const productIsInCart = cartItemsInStorage?.find(
      (item) => item.id === product.id
    )

    if (productIsInCart) {
      if (productIsInCart.quantity === 1) {
        const productsWithoutProduct = cartItemsInStorage.filter(
          (item) => item.id !== product.id
        )

        setStorageItem('cart', productsWithoutProduct)
        return
      }

      const productWithCorrectQuantity = cartItemsInStorage.map((item) =>
        item.id === product.id
          ? { ...productIsInCart, quantity: (productIsInCart.quantity -= 1) }
          : item
      )
      setStorageItem('cart', productWithCorrectQuantity)
    } else {
      setStorageItem('cart', [...cartItemsInStorage, product])
    }
  }

  function clearCart() {
    removeStorageItem('cart')
    setCartItems([] as CartItem[])
  }

  return (
    <CartContext.Provider
      value={{
        cartItems: products || [],
        quantity: cartItems.length || 0,
        total: total || 0,
        addProductToCart,
        handleIncrementProductInCart,
        handleDecrementProductInCart,
        clearCart,
        isLoading
      }}
    >
      {children}
    </CartContext.Provider>
  )
}

export function useCart() {
  return useContext(CartContext)
}
