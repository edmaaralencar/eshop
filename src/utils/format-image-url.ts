export function formatImageUrl(image: string) {
  return `/products/${image}`
}
