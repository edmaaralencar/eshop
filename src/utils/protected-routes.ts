import { GetServerSidePropsContext } from 'next'
// import { unstable_getServerSession } from 'next-auth'
import { getSession } from 'next-auth/react'
// import { options } from 'pages/api/auth/[...nextauth].api'

async function protectedRoutes(context: GetServerSidePropsContext) {
  // const session = await unstable_getServerSession(
  //   context.req,
  //   context.res,
  //   options
  // )

  // if (!session) {
  //   return {
  //     redirect: {
  //       destination: `/sign-in?callbackUrl=${context.resolvedUrl}`,
  //       permanent: false
  //     }
  //   }
  // }

  const session = await getSession(context)

  if (!session) {
    context.res.setHeader(
      'Location',
      `/sign-in?callbackUrl=${context.resolvedUrl}`
    )
    context.res.statusCode = 302
  }

  return session
}

export default protectedRoutes
