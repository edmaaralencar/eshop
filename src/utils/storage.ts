const APP_KEY = '@eshop'

export function getStorageItem(key: string) {
  if (typeof window === 'undefined') return

  const data = window.localStorage.getItem(`${APP_KEY}.${key}`)
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  return JSON.parse(data!)
}

export function setStorageItem(key: string, value: any[]) {
  if (typeof window === 'undefined') return

  const data = JSON.stringify(value)
  return window.localStorage.setItem(`${APP_KEY}.${key}`, data)
}

export function removeStorageItem(key: string) {
  if (typeof window === 'undefined') return

  return window.localStorage.removeItem(`${APP_KEY}.${key}`)
}
