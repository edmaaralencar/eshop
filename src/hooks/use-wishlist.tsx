import { useSession } from 'next-auth/react'
import {
  createContext,
  ReactNode,
  useContext,
  useEffect,
  useState
} from 'react'
import { IProduct } from '../@types/product'
import { getStorageItem, setStorageItem } from 'utils/storage'

type WishlistProviderProps = {
  children: ReactNode
}

type WishlistContextData = {
  items: IProduct[]
  addOrRemoveProductOfWishlist: (product: IProduct) => void
  isProductInWishlist: (id: string) => boolean
}

export const WishlistContext = createContext({} as WishlistContextData)

export function WishlistProvider({ children }: WishlistProviderProps) {
  const [items, setItems] = useState<IProduct[]>([] as IProduct[])
  const { data } = useSession()

  useEffect(() => {
    const wishlistItemsInStorage = getStorageItem('wishlist')

    if (wishlistItemsInStorage) {
      if (data?.user) {
        setItems(wishlistItemsInStorage)
      } else {
        setItems([] as IProduct[])
      }
    }
  }, [data?.user])

  function isProductInWishlist(id: string) {
    return items.some((item) => item.id === id)
  }

  function addOrRemoveProductOfWishlist(product: IProduct) {
    if (isProductInWishlist(product.id)) {
      setItems((prevState) =>
        prevState.filter((item) => item.id !== product.id)
      )
      setStorageItem(
        'wishlist',
        items.filter((item) => item.id !== product.id)
      )
    } else {
      setStorageItem('wishlist', [...items, product])
      setItems((prevState) => [...prevState, product])
    }
  }

  return (
    <WishlistContext.Provider
      value={{
        items,
        isProductInWishlist,
        addOrRemoveProductOfWishlist
      }}
    >
      {children}
    </WishlistContext.Provider>
  )
}

export function useWishlist() {
  return useContext(WishlistContext)
}
