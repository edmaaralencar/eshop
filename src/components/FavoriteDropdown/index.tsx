import Button from 'components/Button'
import Hr from 'components/Hr'
import IconButton from 'components/IconButton'
import { useWishlist } from 'hooks/use-wishlist'
import Image from 'next/image'
import Link from 'next/link'
import { Fragment } from 'react'
import { FiHeart } from 'react-icons/fi'
import { formatImageUrl } from 'utils/format-image-url'
import { formatMoney } from 'utils/format-money'

import * as S from './styles'

function FavoriteDropdown() {
  const { items, addOrRemoveProductOfWishlist, productIsInWishlist } =
    useWishlist()

  return (
    <S.Wrapper>
      <S.Content>
        <S.Arrow />

        <S.FavoritesList>
          {items?.length < 1 ? (
            <S.EmptyList>
              <p>Nenhum item salvo.</p>
            </S.EmptyList>
          ) : (
            <>
              {items?.map((product) => (
                <Fragment key={product.id}>
                  <S.Favorite>
                    <S.FavoriteContainer>
                      <Image
                        alt={product.name}
                        src={formatImageUrl(product.images[0].image_url)}
                        width={100}
                        height={80}
                      />
                      <div>
                        <span>{product.name}</span>
                        <strong>{formatMoney(product.price)}</strong>
                      </div>
                    </S.FavoriteContainer>

                    <S.ButtonContainer>
                      <IconButton
                        onClick={() => addOrRemoveProductOfWishlist(product)}
                        icon={
                          productIsInWishlist(product.id) ? (
                            <FiHeart size={18} fill="#6C6CFF" color="#6C6CFF" />
                          ) : (
                            <FiHeart size={18} color="#6C6CFF" />
                          )
                        }
                      />
                    </S.ButtonContainer>

                    <Link href={`/products/${product.id}`} passHref>
                      <Button as="a" variant="secondary" fullWidth size="small">
                        Ver mais
                      </Button>
                    </Link>
                  </S.Favorite>
                  <Hr />
                </Fragment>
              ))}
            </>
          )}
        </S.FavoritesList>
      </S.Content>
    </S.Wrapper>
  )
}

export default FavoriteDropdown
