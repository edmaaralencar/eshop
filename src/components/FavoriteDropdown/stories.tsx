import { StoryObj, Meta } from '@storybook/react'
import * as HoverCard from '@radix-ui/react-hover-card'
import FavoriteDropdown from '.'

export default {
  title: 'FavoriteDropdown',
  component: FavoriteDropdown,
  decorators: [
    (Story) => (
      <HoverCard.Root openDelay={0} closeDelay={0}>
        <Story />
        <HoverCard.Trigger asChild>
          <button>Abrir wishlist</button>
        </HoverCard.Trigger>
      </HoverCard.Root>
    )
  ]
} as Meta

export const Default: StoryObj = {}
