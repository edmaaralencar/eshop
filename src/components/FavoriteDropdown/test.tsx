import { render, screen } from '@testing-library/react'

import * as HoverCard from '@radix-ui/react-hover-card'
import { WishlistContext } from 'hooks/use-wishlist'

import FavoriteDropdown from '.'
import { renderWithTheme } from 'utils/test-helper'

global.ResizeObserver = jest.fn().mockImplementation(() => ({
  observe: jest.fn(),
  unobserve: jest.fn(),
  disconnect: jest.fn()
}))

const product = {
  id: '1231',
  name: 'Air Max 90',
  description: 'oiad',
  price: 300,
  images: [
    {
      image_url: '/produto.png',
      id: '12'
    }
  ]
}

describe('<FavoriteDropdown />', () => {
  it('should render the dropdown with products', () => {
    const { container } = renderWithTheme(
      <HoverCard.Root open>
        <WishlistContext.Provider
          value={{
            items: [product],
            addOrRemoveProductOfWishlist: jest.fn(),
            productIsInWishlist: jest.fn()
          }}
        >
          <FavoriteDropdown />
        </WishlistContext.Provider>
      </HoverCard.Root>
    )

    expect(screen.getByRole('link', { name: /ver mais/i })).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render the dropdown with products', () => {
    renderWithTheme(
      <HoverCard.Root open>
        <WishlistContext.Provider
          value={{
            items: [],
            addOrRemoveProductOfWishlist: jest.fn(),
            productIsInWishlist: jest.fn()
          }}
        >
          <FavoriteDropdown />
        </WishlistContext.Provider>
      </HoverCard.Root>
    )

    expect(screen.getByText(/nenhum item salvo./i)).toBeInTheDocument()
  })
})
