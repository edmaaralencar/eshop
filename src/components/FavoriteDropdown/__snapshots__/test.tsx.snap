// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FavoriteDropdown /> should render the dropdown with products 1`] = `null`;
